In [ ]:
import os,shutil

**Importing dataset from Kaggle**

In [ ]:
os.environ['KAGGLE_USERNAME']='rohanazad'
os.environ['KAGGLE_KEY']='f3cf3c594532e397b78564e4bbe88ec6'

In [ ]:
!kaggle competitions download -c ashrae-energy-prediction

 87% 77.0M/88.4M [00:01<00:00, 44.5MB/s]
100% 88.4M/88.4M [00:01<00:00, 76.1MB/s]
  0% 0.00/1.27M [00:00<?, ?B/s]
100% 1.27M/1.27M [00:00<00:00, 183MB/s]
 93% 155M/167M [00:02<00:00, 45.5MB/s]
100% 167M/167M [00:02<00:00, 64.2MB/s]
 88% 105M/120M [00:02<00:00, 29.3MB/s] 
100% 120M/120M [00:02<00:00, 46.1MB/s]
  0% 0.00/44.5k [00:00<?, ?B/s]
100% 44.5k/44.5k [00:00<00:00, 38.9MB/s]
  0% 0.00/2.53M [00:00<?, ?B/s]
100% 2.53M/2.53M [00:00<00:00, 83.3MB/s]


In [ ]:
import pandas as pd

In [ ]:
import zipfile

unzipped_train=zipfile.ZipFile('train.csv.zip')
unzipped_train.extractall()
unzipped_train.close()

unzipped_test=zipfile.ZipFile('test.csv.zip')
unzipped_test.extractall()
unzipped_test.close()

In [ ]:
unzipped_weather_train=zipfile.ZipFile('weather_train.csv.zip')
unzipped_weather_train.extractall()
unzipped_weather_train.close()

unzipped_weather_test=zipfile.ZipFile('weather_test.csv.zip')
unzipped_weather_test.extractall()
unzipped_weather_test.close()

In [ ]:
# weather_train=pd.read_csv('weather_train.csv')
# weather_test=pd.read_csv('weather_test.csv')
# building_met=pd.read_csv('building_metadata.csv')

In [ ]:
# total=weather_train.isnull().sum().sort_values(ascending=False)
# percent=(weather_train.isnull().sum()/weather_train.isnull().count()).sort_values(ascending=False)
# missing_values=pd.concat([total,percent],axis=1,keys=['total','percent'])
# missing_values.head(20)

In [ ]:
# total=building_met.isnull().sum().sort_values(ascending=False)
# percent=(building_met.isnull().sum()/building_met.isnull().count()).sort_values(ascending=False)
# missing_values=pd.concat([total,percent],axis=1,keys=['total','percent'])
# missing_values.head(20)

In [ ]:
train=pd.read_csv('train.csv')
test=pd.read_csv('test.csv')
weather_train_dir='/content/drive/My Drive/weather_train3.csv'
weather_train=pd.read_csv(weather_train_dir)
weather_test_dir='/content/drive/My Drive/weather_test2.csv'
weather_test=pd.read_csv(weather_test_dir)
building_met_dir='/content/drive/My Drive/building_metadata_save.csv'
building_met=pd.read_csv(building_met_dir)

In [ ]:
# total=train.isnull().sum().sort_values(ascending=False)
# percent=(train.isnull().sum()/train.isnull().count()).sort_values(ascending=False)
# missing_values=pd.concat([total,percent],axis=1,keys=['total','percent'])
# missing_values.head(20)

In [ ]:
# train.head()

In [ ]:
print('shape of train:',train.shape)
print('shape of test:',test.shape)
print('shape of weather_train:',weather_train.shape)
print('shape of weather_test:',weather_test.shape)
print('shape of building_metadata:',building_met.shape)

shape of train: (20216100, 4)
shape of test: (41697600, 4)
shape of weather_train: (139788, 12)
shape of weather_test: (277243, 12)
shape of building_metadata: (1449, 6)


**timestamp stores data as strings in a particular format. Year,month,day,time**

**Apply to_datetime to convert string to object**

In [ ]:
train['timestamp']=pd.to_datetime(train['timestamp'])
test['timestamp']=pd.to_datetime(test['timestamp'])
weather_train['timestamp']=pd.to_datetime(weather_train['timestamp'])
weather_test['timestamp']=pd.to_datetime(weather_test['timestamp'])

building_met['primary_use']=building_met['primary_use'].astype('category')

**Merging the three dataset to create train set**

In [ ]:
temp=train[['building_id']]
temp=temp.merge(building_met,on='building_id',how='left')
del temp['building_id']
train=pd.concat([train,temp],axis=1)



In [ ]:
train.head()

building_id  meter  timestamp  ...  square_feet  year_built floor_count
0            0      0 2016-01-01  ...         7432      2008.0           4
1            1      0 2016-01-01  ...         2720      2004.0           4
2            2      0 2016-01-01  ...         5376      1991.0           4
3            3      0 2016-01-01  ...        23685      2002.0           4
4            4      0 2016-01-01  ...       116607      1975.0           4

[5 rows x 9 columns]

In [ ]:
temp=train[['site_id','timestamp']]
temp=temp.merge(weather_train,on=['site_id','timestamp'],how='left')
del temp['site_id'],temp['timestamp']
train=pd.concat([train,temp],axis=1)

del temp,weather_train

In [ ]:
train.head()

building_id  meter  timestamp  ...  Month  Day_date    year
0            0      0 2016-01-01  ...    1.0       1.0  2016.0
1            1      0 2016-01-01  ...    1.0       1.0  2016.0
2            2      0 2016-01-01  ...    1.0       1.0  2016.0
3            3      0 2016-01-01  ...    1.0       1.0  2016.0
4            4      0 2016-01-01  ...    1.0       1.0  2016.0

[5 rows x 19 columns]

In [ ]:
train.shape

(20216100, 19)

In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20216100 entries, 0 to 20216099
Data columns (total 19 columns):
 #   Column              Dtype         
---  ------              -----         
 0   building_id         int64         
 1   meter               int64         
 2   timestamp           datetime64[ns]
 3   meter_reading       float64       
 4   site_id             int64         
 5   primary_use         category      
 6   square_feet         int64         
 7   year_built          float64       
 8   floor_count         int64         
 9   air_temperature     float64       
 10  cloud_coverage      float64       
 11  dew_temperature     float64       
 12  precip_depth_1_hr   float64       
 13  sea_level_pressure  float64       
 14  wind_direction      float64       
 15  wind_speed          float64       
 16  Month               float64       
 17  Day_date            float64       
 18  year                float64       
dtypes: category(1), datetime64[ns](1), float

**Creating Hour column**

In [ ]:
import numpy as np

train['hour_datetime']=train['timestamp'].dt.hour.astype(np.int8)

del train['timestamp']

In [ ]:
train.head()

building_id  meter  meter_reading  ...  Day_date    year  hour_datetime
0            0      0            0.0  ...       1.0  2016.0              0
1            1      0            0.0  ...       1.0  2016.0              0
2            2      0            0.0  ...       1.0  2016.0              0
3            3      0            0.0  ...       1.0  2016.0              0
4            4      0            0.0  ...       1.0  2016.0              0

[5 rows x 19 columns]

**Looking for missing values**

In [ ]:
total=train.isnull().sum().sort_values(ascending=False)
percent=(train.isnull().sum()/train.isnull().count()).sort_values(ascending=False)
missing_values=pd.concat([total,percent],axis=1,keys=['total','percent'])
missing_values.head(20)

total  percent
cloud_coverage      82890   0.0041
Day_date            82890   0.0041
Month               82890   0.0041
wind_speed          82890   0.0041
wind_direction      82890   0.0041
sea_level_pressure  82890   0.0041
precip_depth_1_hr   82890   0.0041
dew_temperature     82890   0.0041
year                82890   0.0041
air_temperature     82890   0.0041
primary_use             0   0.0000
meter                   0   0.0000
meter_reading           0   0.0000
site_id                 0   0.0000
hour_datetime           0   0.0000
square_feet             0   0.0000
year_built              0   0.0000
floor_count             0   0.0000
building_id             0   0.0000

In [ ]:
train=train.dropna()

In [ ]:
train.describe()

building_id         meter  ...        year  hour_datetime
count  2.013321e+07  2.013321e+07  ...  20133210.0   2.013321e+07
mean   7.974179e+02  6.617699e-01  ...      2016.0   1.150934e+01
std    4.264339e+02  9.310602e-01  ...         0.0   6.923804e+00
min    0.000000e+00  0.000000e+00  ...      2016.0   0.000000e+00
25%    3.910000e+02  0.000000e+00  ...      2016.0   6.000000e+00
50%    8.940000e+02  0.000000e+00  ...      2016.0   1.200000e+01
75%    1.176000e+03  1.000000e+00  ...      2016.0   1.800000e+01
max    1.448000e+03  3.000000e+00  ...      2016.0   2.300000e+01

[8 rows x 18 columns]

In [ ]:
total=train.isnull().sum().sort_values(ascending=False)
percent=(train.isnull().sum()/train.isnull().count()).sort_values(ascending=False)
missing_values=pd.concat([total,percent],axis=1,keys=['total','percent'])
missing_values.head(20)

total  percent
hour_datetime           0      0.0
air_temperature         0      0.0
meter                   0      0.0
meter_reading           0      0.0
site_id                 0      0.0
primary_use             0      0.0
square_feet             0      0.0
year_built              0      0.0
floor_count             0      0.0
cloud_coverage          0      0.0
year                    0      0.0
dew_temperature         0      0.0
precip_depth_1_hr       0      0.0
sea_level_pressure      0      0.0
wind_direction          0      0.0
wind_speed              0      0.0
Month                   0      0.0
Day_date                0      0.0
building_id             0      0.0

**Before applying pipeline transformation, hot water dataset has been separated**

In [ ]:
dataset_hot_water=train.loc[(train['meter']==3)]

In [ ]:
dataset_hot_water.head()

building_id  meter  meter_reading  ...  Day_date    year  hour_datetime
105          106      3          0.000  ...       1.0  2016.0              0
109          109      3          0.000  ...       1.0  2016.0              0
113          112      3         96.978  ...       1.0  2016.0              0
115          113      3         19.597  ...       1.0  2016.0              0
117          114      3        100.000  ...       1.0  2016.0              0

[5 rows x 19 columns]

In [ ]:
dataset_hot_water.shape

(1261500, 19)

In [ ]:
drop_cols=['meter_reading']

In [ ]:
training_hot_water=dataset_hot_water.drop(drop_cols,axis=1)

In [ ]:
target_hot_water=dataset_hot_water['meter_reading']

In [ ]:
training_hot_water.shape

(1261500, 18)

In [ ]:
target_hot_water.shape

(1261500,)

In [ ]:
training_hot_water['primary_use'].value_counts()

Education                        606702
Office                           289492
Entertainment/public assembly    138877
Lodging/residential               95829
Public services                   78943
Healthcare                        25327
Food sales and service            17557
Technology/science                 8773
Warehouse/storage                     0
Utility                               0
Services                              0
Retail                                0
Religious worship                     0
Parking                               0
Other                                 0
Manufacturing/industrial              0
Name: primary_use, dtype: int64

**Apply pipeline transformation to training_chilled_water**

In [ ]:
num_cols=['square_feet','floor_count','air_temperature',
                 'cloud_coverage','dew_temperature','precip_depth_1_hr',
                 'sea_level_pressure','wind_direction','wind_speed']
cat_cols=['primary_use']

In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder 

num_attribs=list(num_cols)
cat_attribs=['primary_use']

num_pipeline=Pipeline([
    ('std_scaler',StandardScaler())
])

full_pipeline=ColumnTransformer([
    ('num',num_pipeline,num_attribs),
    ('cat',OneHotEncoder(),cat_attribs)
])

training_prepared=full_pipeline.fit_transform(training_hot_water)

In [ ]:
x_train=training_prepared[:750811]
x_val=training_prepared[750811:]
y_train=target_hot_water[:750811]
y_val=target_hot_water[750811:]

In [ ]:
print('shape of x_train:',x_train.shape)
print('shape of x_val:',x_val.shape)
print('shape of y_train:',y_train.shape)
print('shape of y_val:',y_val.shape)

shape of x_train: (750811, 17)
shape of x_val: (510689, 17)
shape of y_train: (750811,)
shape of y_val: (510689,)


In [ ]:
x_train[0]

array([-1.08331499, -0.13617522, -1.15213176, -0.61984805, -0.61984805,
       -0.20624077,  0.88110768,  0.58337957, -0.01330586,  1.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ])

**Working with test set**

In [ ]:
test.head()

row_id  building_id  meter  timestamp
0       0            0      0 2017-01-01
1       1            1      0 2017-01-01
2       2            2      0 2017-01-01
3       3            3      0 2017-01-01
4       4            4      0 2017-01-01

In [ ]:
test.shape

(41697600, 4)

In [ ]:
import pandas as pd

temp=test[['building_id']]
temp=temp.merge(building_met,on='building_id',how='left')
del temp['building_id']
test=pd.concat([test,temp],axis=1)

In [ ]:
test.head()

row_id  building_id  meter  ... square_feet  year_built floor_count
0       0            0      0  ...        7432      2008.0           4
1       1            1      0  ...        2720      2004.0           4
2       2            2      0  ...        5376      1991.0           4
3       3            3      0  ...       23685      2002.0           4
4       4            4      0  ...      116607      1975.0           4

[5 rows x 9 columns]

In [ ]:
temp=test[['site_id','timestamp']]
temp=temp.merge(weather_test,on=['site_id','timestamp'],how='left')
del temp['timestamp'],temp['site_id']
test=pd.concat([test,temp],axis=1)

del temp,weather_test

In [ ]:
test.head()

row_id  building_id  meter  timestamp  ...  wind_speed Month  Day_date    year
0       0            0      0 2017-01-01  ...         3.6   1.0       1.0  2017.0
1       1            1      0 2017-01-01  ...         3.6   1.0       1.0  2017.0
2       2            2      0 2017-01-01  ...         3.6   1.0       1.0  2017.0
3       3            3      0 2017-01-01  ...         3.6   1.0       1.0  2017.0
4       4            4      0 2017-01-01  ...         3.6   1.0       1.0  2017.0

[5 rows x 19 columns]

In [ ]:
import numpy as np

test['hour_datetime']=test['timestamp'].dt.hour.astype(np.int8)

del test['timestamp']

In [ ]:
total=test.isnull().sum().sort_values(ascending=False)
percent=(test.isnull().sum()/test.isnull().count()).sort_values(ascending=False)
missing_values=pd.concat([total,percent],axis=1,keys=['total','percent'])
missing_values.head(20)

total   percent
cloud_coverage      199029  0.004773
Day_date            199029  0.004773
Month               199029  0.004773
wind_speed          199029  0.004773
wind_direction      199029  0.004773
sea_level_pressure  199029  0.004773
precip_depth_1_hr   199029  0.004773
dew_temperature     199029  0.004773
year                199029  0.004773
air_temperature     199029  0.004773
primary_use              0  0.000000
building_id              0  0.000000
meter                    0  0.000000
site_id                  0  0.000000
hour_datetime            0  0.000000
square_feet              0  0.000000
year_built               0  0.000000
floor_count              0  0.000000
row_id                   0  0.000000

In [ ]:
test.describe()

row_id   building_id  ...          year  hour_datetime
count  4.169760e+07  4.169760e+07  ...  4.149857e+07   4.169760e+07
mean   2.084880e+07  8.075824e+02  ...  2.017500e+03   1.150000e+01
std    1.203706e+07  4.297680e+02  ...  5.000000e-01   6.922187e+00
min    0.000000e+00  0.000000e+00  ...  2.017000e+03   0.000000e+00
25%    1.042440e+07  4.047500e+02  ...  2.017000e+03   5.750000e+00
50%    2.084880e+07  9.000000e+02  ...  2.017000e+03   1.150000e+01
75%    3.127320e+07  1.194250e+03  ...  2.018000e+03   1.725000e+01
max    4.169760e+07  1.448000e+03  ...  2.018000e+03   2.300000e+01

[8 rows x 18 columns]

In [ ]:
test_hot_water= test.loc[(test['meter']==3)]

In [ ]:
test_hot_water.shape

(2540400, 19)

In [ ]:
test_drop_cols=['row_id']

In [ ]:
testing=test_hot_water.drop(test_drop_cols,axis=1)

In [ ]:
testing.shape

(2540400, 18)

In [ ]:
testing_prepared=full_pipeline.transform(testing)

In [ ]:
testing_prepared.shape

(2540400, 17)

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from math import sqrt

lin_reg=LinearRegression()
lin_reg.fit(x_train,y_train)
lin_predict=lin_reg.predict(x_val)
lin_mse=mean_squared_error(lin_predict,y_val)
lin_rmse=sqrt(lin_mse)
lin_rmse

2930.9912070032638

Linear regression RMSE 2930.9912070032638

**Applying 2 degree polynomial**

In [ ]:
from sklearn.preprocessing import PolynomialFeatures

poly_features=PolynomialFeatures(degree=2,include_bias=False)
x_poly=poly_features.fit_transform(x_train)
lin_reg=LinearRegression()
lin_reg.fit(x_poly,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
x_val_poly=poly_features.transform(x_val)
predict=lin_reg.predict(x_val_poly)
degree2_mse=mean_squared_error(predict,y_val)
degree2_rmse=sqrt(degree2_mse)
degree2_rmse

2486.5616840162443

2nd degree polynomial RMSE 2486.5616840162443

**Using grid search with Ridge regression**

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Ridge
param_grid=[
            {'alpha':[0.001,0.003,0.01,0.03,0.1,0.3,1,3]}
]

ridge_reg=Ridge()

grid_search=GridSearchCV(ridge_reg,param_grid,cv=5,
                         scoring='neg_mean_squared_error',
                         return_train_score=True)

grid_search.fit(x_train,y_train)

GridSearchCV(cv=5, error_score=nan,
             estimator=Ridge(alpha=1.0, copy_X=True, fit_intercept=True,
                             max_iter=None, normalize=False, random_state=None,
                             solver='auto', tol=0.001),
             iid='deprecated', n_jobs=None,
             param_grid=[{'alpha': [0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1, 3]}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
             scoring='neg_mean_squared_error', verbose=0)

In [ ]:
grid_search.best_params_

{'alpha': 3}

Ridge regression alpha=3


**Ridge regression with alpha=3**

In [ ]:
from sklearn.metrics import mean_squared_error

ridge_reg=Ridge(alpha=0.001,solver='cholesky')
ridge_reg.fit(x_train,y_train)
ridge_pred=ridge_reg.predict(x_val)
ridge_mse=mean_squared_error(ridge_pred,y_val)
ridge_rmse=np.sqrt(ridge_mse)
ridge_rmse

2930.991207133157

Ridge regression RMSE 2930.991207133157 

**Using grid search with Lasso regression**

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Lasso

param_grid=[
            {'alpha':[0.001,0.003,0.01,0.03,0.1,0.3,1,3]}
]

lasso_reg=Lasso()

grid_search=GridSearchCV(lasso_reg,param_grid,cv=5,
                         scoring='neg_mean_squared_error',
                         return_train_score=True)

grid_search.fit(x_train,y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_coordinate_descent.py:476: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 269324296898.9441, tolerance: 257242302.19205338
  positive)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_coordinate_descent.py:476: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 44176121524.98633, tolerance: 215381640.22858506
  positive)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_coordinate_descent.py:476: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 62550590294.43408, tolerance: 310948226.5984097
  positive)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_coordinate_descent.py:476: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 6476341793

GridSearchCV(cv=5, error_score=nan,
             estimator=Lasso(alpha=1.0, copy_X=True, fit_intercept=True,
                             max_iter=1000, normalize=False, positive=False,
                             precompute=False, random_state=None,
                             selection='cyclic', tol=0.0001, warm_start=False),
             iid='deprecated', n_jobs=None,
             param_grid=[{'alpha': [0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1, 3]}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
             scoring='neg_mean_squared_error', verbose=0)

In [ ]:
grid_search.best_params_

{'alpha': 3}

Lasso regression alpha= 3

**Lasso regression with alpha= 3**

In [ ]:
from sklearn.linear_model import Lasso

lasso_reg=Lasso(alpha=3)
lasso_reg.fit(x_train,y_train)
lasso_pred=lasso_reg.predict(x_val)
lasso_mse=mean_squared_error(lasso_pred,y_val)
lasso_rmse=np.sqrt(lasso_mse)
lasso_rmse

2931.682913534796

Lasso regression RMSE 2931.682913534796

**Grid search with elastic net**

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import ElasticNet

param_grid=[
            {'alpha':[0.001,0.003,0.01,0.03,0.1,0.3,1,3]}
]

elastic_reg=ElasticNet()

grid_search=GridSearchCV(elastic_reg,param_grid,cv=5,
                         scoring='neg_mean_squared_error',
                         return_train_score=True)

grid_search.fit(x_train,y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_coordinate_descent.py:476: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1137131953586.7896, tolerance: 257242302.19205338
  positive)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_coordinate_descent.py:476: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 948422060001.4845, tolerance: 215381640.22858506
  positive)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_coordinate_descent.py:476: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1379295069413.107, tolerance: 310948226.5984097
  positive)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_coordinate_descent.py:476: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 144680880

GridSearchCV(cv=5, error_score=nan,
             estimator=ElasticNet(alpha=1.0, copy_X=True, fit_intercept=True,
                                  l1_ratio=0.5, max_iter=1000, normalize=False,
                                  positive=False, precompute=False,
                                  random_state=None, selection='cyclic',
                                  tol=0.0001, warm_start=False),
             iid='deprecated', n_jobs=None,
             param_grid=[{'alpha': [0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1, 3]}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
             scoring='neg_mean_squared_error', verbose=0)

In [ ]:
grid_search.best_params_

{'alpha': 0.1}

Elastic net alpha=0.1


**Applying elastic net with alpha=0.1**

In [ ]:
from sklearn.linear_model import ElasticNet
from sklearn.metrics import mean_squared_error


elastic_reg=ElasticNet(alpha=0.1)
elastic_reg.fit(x_train,y_train)
elastic_pred=elastic_reg.predict(x_val)
elastic_mse=mean_squared_error(elastic_pred,y_val)
elastic_rmse=np.sqrt(elastic_mse)
elastic_rmse

2935.439969289137

Elastic net RMSE= 2935.439969289137

**Grid search to get the optimal learning rate**

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import SGDRegressor

param_grid=[
            {'eta0':[0.001,0.003,0.01,0.03,0.1,0.3,1,3]}
]

sgd_reg=SGDRegressor(max_iter=1000,tol=1e-3,penalty='l2')

grid_search=GridSearchCV(sgd_reg,param_grid,cv=5,
                         scoring='neg_mean_squared_error',
                         return_train_score=True)
grid_search.fit(x_train,y_train)

GridSearchCV(cv=5, error_score=nan,
             estimator=SGDRegressor(alpha=0.0001, average=False,
                                    early_stopping=False, epsilon=0.1,
                                    eta0=0.01, fit_intercept=True,
                                    l1_ratio=0.15, learning_rate='invscaling',
                                    loss='squared_loss', max_iter=1000,
                                    n_iter_no_change=5, penalty='l2',
                                    power_t=0.25, random_state=None,
                                    shuffle=True, tol=0.001,
                                    validation_fraction=0.1, verbose=0,
                                    warm_start=False),
             iid='deprecated', n_jobs=None,
             param_grid=[{'eta0': [0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1, 3]}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
             scoring='neg_mean_squared_error', verbose=0)

In [ ]:
grid_search.best_params_

{'eta0': 0.03}

**Stochastic gradient descent with eta0=0.03**

In [ ]:
from sklearn.linear_model import SGDRegressor
from sklearn. metrics import mean_squared_error

sgd_reg=SGDRegressor(max_iter=200,tol=1e-3,eta0=0.03)
sgd_reg.fit(x_train,y_train)
sgd_pred=sgd_reg.predict(x_val)
sgd_mse=mean_squared_error(sgd_pred,y_val)
sgd_rmse=np.sqrt(sgd_mse)
sgd_rmse

2940.1910539915416

Stochastic gradient descent RMSE with eta0=0.03 is 2940.1910539915416

**Support Vector Regressor**

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.svm import LinearSVR

param_grid = [
              {'epsilon':[0.003,0.001,0.03,0.01,0.3,0.1,1,3,10,20,50,100,400,800,1000]}
]

svm_reg = LinearSVR()
grid_search = GridSearchCV(svm_reg,param_grid,
                           scoring ='neg_mean_squared_error',
                           return_train_score =True)
grid_search.fit(x_train,y_train)

GridSearchCV(cv=None, error_score=nan,
             estimator=LinearSVR(C=1.0, dual=True, epsilon=0.0,
                                 fit_intercept=True, intercept_scaling=1.0,
                                 loss='epsilon_insensitive', max_iter=1000,
                                 random_state=None, tol=0.0001, verbose=0),
             iid='deprecated', n_jobs=None,
             param_grid=[{'epsilon': [0.003, 0.001, 0.03, 0.01, 0.3, 0.1, 1, 3,
                                      10, 20, 50, 100, 400, 800, 1000]}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
             scoring='neg_mean_squared_error', verbose=0)

In [ ]:
grid_search.best_params_

{'epsilon': 400}

**from grid search epsilon = 400. LinearSVR with epsilon = 400**

In [ ]:
from sklearn.svm import LinearSVR
from sklearn.metrics import mean_squared_error
from math import sqrt

svm_reg = LinearSVR(epsilon =400)
svm_reg.fit(x_train,y_train)
svm_reg_pred = svm_reg.predict(x_val)
svm_reg_error = mean_squared_error(svm_reg_pred,y_val)
svm_reg_rmse = sqrt(svm_reg_error)
svm_reg_rmse

2976.1475145440604

LinearSVR with epsilon 1000 RMSE = 2976.1475145440604

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVR

param_grid = [
              {'epsilon':[10,20,50,100,400,800,1000]},
              {'degree':[2,3,4,5]},
              {'C':[0.003,0.001,0.03,0.01]} #,20,50,100,400,800,1000
]

svm_poly_reg = SVR(kernel = 'poly')

grid_search = GridSearchCV(svm_poly_reg,param_grid,
                           scoring ='neg_mean_squared_error',
                           return_train_score =True)

grid_search.fit(x_train,y_train)

In [ ]:
grid_search.bet_params_

In [ ]:
from sklearn.svm import SVR
from sklearn.metrics mean_squared_error
from math import sqrt

svm_poly_reg = SVR(kernel ='poly', degree = '', C ='', epislon ='')
svm_poly_reg.fit(x_train,y_train)
svm_poly_pred = svm_poly_reg.pred(x_val)
svm_poly_mse=mean_squared_error(svm_poy_pred,y_val)
svm_poly_rmse = sqrt(svm_poly_mse)
svm_poy_rmse

**Decision tree**

**The plan was to let decision tree fit however it wanted and let it overfit, make the figure using export_graphviz and then reducing overfitting**

In [ ]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error

tree_reg=DecisionTreeRegressor()
tree_reg.fit(x_train,y_train)
tree_pred=tree_reg.predict(x_val)
tree_mse=mean_squared_error(tree_pred,y_val)
tree_rmse=np.sqrt(tree_mse)
tree_rmse

3071.9510853810343

Decision tree RMSE = 3071.9510853810343

In [ ]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error

tree_reg=DecisionTreeRegressor(max_depth =10, max_leaf_nodes=10)
tree_reg.fit(x_train,y_train)
tree_pred=tree_reg.predict(x_val)
tree_mse=mean_squared_error(tree_pred,y_val)
tree_rmse=np.sqrt(tree_mse)
tree_rmse

2489.6482968306777

**Ensemble**

In [ ]:
from sklearn.ensemble import VotingRegressor
from sklearn.svm import LinearSVR
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.linear_model import ElasticNet
from sklearn.linear_model import SGDRegressor
from sklearn.svm import LinearSVR
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor

lin_reg=LinearRegression()
#ridge_reg=Ridge(alpha=0.001,solver='cholesky')
#lasso_reg=Lasso(alpha=0.001)
#elastic_reg=ElasticNet(alpha=0.003)
#sgd_reg=SGDRegressor(max_iter=200,tol=1e-3,eta0=0.03)
#svm_reg = LinearSVR(epsilon =400)
tree_reg=DecisionTreeRegressor(max_depth =10, max_leaf_nodes=10)

voting_reg=VotingRegressor(
    estimators=[('lr',lin_reg), 
                ('tree',tree_reg)]
)

voting_reg.fit(x_train,y_train)

VotingRegressor(estimators=[('lr',
                             LinearRegression(copy_X=True, fit_intercept=True,
                                              n_jobs=None, normalize=False)),
                            ('tree',
                             DecisionTreeRegressor(ccp_alpha=0.0,
                                                   criterion='mse',
                                                   max_depth=10,
                                                   max_features=None,
                                                   max_leaf_nodes=10,
                                                   min_impurity_decrease=0.0,
                                                   min_impurity_split=None,
                                                   min_samples_leaf=1,
                                                   min_samples_split=2,
                                                   min_weight_fraction_leaf=0.0,
                                                   pre

In [ ]:
from sklearn.metrics import mean_squared_error

voting_reg_pred=voting_reg.predict(x_val)
voting_reg_mse=mean_squared_error(voting_reg_pred,y_val)
voting_reg_rmse=np.sqrt(voting_reg_mse)
voting_reg_rmse

2669.26409222902

Voting regressor RMSE 2669.26409222902

**BaggingRegressor**

In [ ]:
from sklearn.ensemble import BaggingRegressor
from sklearn.tree import DecisionTreeRegressor

bag_reg= BaggingRegressor(
    DecisionTreeRegressor(max_depth =10, max_leaf_nodes=10),n_estimators =1000,
    max_samples=2000, bootstrap =True, n_jobs=-1 
)


In [ ]:
from math import sqrt

bag_reg.fit(x_train,y_train)
bag_pred= bag_reg.predict(x_val)
bag_mse=mean_squared_error(y_val,bag_pred)
bag_rmse =sqrt(bag_mse)

In [ ]:
bag_rmse

2531.0452042743564

Bagging regressor RMSE = 2531.0452042743564

**RandomFOrest**

**Gardient boosting**

In [ ]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error

gbrt = GradientBoostingRegressor(max_depth =10, max_leaf_nodes=10,n_estimators=500,min_samples_leaf=1000,max_features =9,  random_state=42)
gbrt.fit(x_train, y_train)

errors = [mean_squared_error(y_val, y_pred)
          for y_pred in gbrt.staged_predict(x_val)]
bst_n_estimators = np.argmin(errors) + 1

gbrt_best = GradientBoostingRegressor(max_depth=2, n_estimators=bst_n_estimators, random_state=42)
gbrt_best.fit(x_train, y_train)

GradientBoostingRegressor(alpha=0.9, ccp_alpha=0.0, criterion='friedman_mse',
                          init=None, learning_rate=0.1, loss='ls', max_depth=2,
                          max_features=None, max_leaf_nodes=None,
                          min_impurity_decrease=0.0, min_impurity_split=None,
                          min_samples_leaf=1, min_samples_split=2,
                          min_weight_fraction_leaf=0.0, n_estimators=105,
                          n_iter_no_change=None, presort='deprecated',
                          random_state=42, subsample=1.0, tol=0.0001,
                          validation_fraction=0.1, verbose=0, warm_start=False)

In [ ]:
gbrt_best_pred=gbrt_best.predict(x_val)
gbrt_best_mse=mean_squared_error(gbrt_best_pred,y_val)
gbrt_best_rmse = sqrt(gbrt_best_mse)
gbrt_best_rmse

2510.493479511271

gradient boosting RMSE with max_depth=10, max_leaf_nodes=10,n_estimators=500,random_state=2510.493479511271///// 

**Making test set predictions**

In [ ]:
tree_reg_test= tree_reg.predict(testing_prepared)

**Saving test set predictions**

In [ ]:
hot_water_preds = pd.DataFrame({'row_id':test_hot_water['row_id'],
                            'meter_reading':tree_reg_test})
hot_water_preds.head()

row_id  meter_reading
2260082  2260082     214.728357
2260086  2260086     214.728357
2260090  2260090     214.728357
2260092  2260092     214.728357
2260094  2260094     214.728357

In [ ]:
file_name='hot_water_preds.csv'
hot_water_preds.to_csv(file_name,index=False)
print('Saved file:'+file_name)

Saved file:hot_water_preds.csv
